### PR TITLE
Overview image

### PR DESCRIPTION
**Pull request by philipwinklerdelmic**

### CODE REVIEW

File: src/odemis/gui/cont/views.py [hunk 1]
```diff
@@ -317,6 +370,32 @@ def on_stage_pos_change(self, p_pos):
...
         self._data_model.stage_history.value = stage_history
 
+        # Update overview image
+        self._update_ovv()
+
+    def _on_current_stream(self, streams):
+        """
+        Called when some VAs affecting the current stream change
+        """
+        # Try to get the current stream
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: You need to unsubscribe from the previous curr_s first (here).

File: src/odemis/gui/cont/views.py [hunk 2]
```diff
@@ -325,31 +404,55 @@ def _on_chamber_state(self, state):
             # sample have probably nothing in common with this new sample
             self._data_model.stage_history.value = self._data_model.stage_history.value[-1:]
 
+    def _update_ovv(self):
+        """ Update the overview image with the currently active stream. """
+        if self._data_model.streams.value:
+            for s in self._data_model.streams.value:
+                if s.is_active.value and s.image.value is not None:
+                    img = s.image.value
+                    if isinstance(s, acqstream.OpticalStream):
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: Use `if...ifelse...else`, to catch also other types of streams (and at least, log that you're not putting this other types into the overview image).

File: src/odemis/gui/cont/views.py [hunk 3]
```diff
@@ -325,31 +404,55 @@ def _on_chamber_state(self, state):
...
+                        opt = img
+                        sem = None
+                    if isinstance(s, acqstream.EMStream):
+                        sem = img
+                        opt = None
+
+                    # Update optical or sem image, depending on which stream is active
+                    if opt is not None:
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: That if block seems redundant with the previous one. Just merge them together, and drop the `opt` & `sem` variables.

File: src/odemis/gui/cont/views.py [hunk 4]
```diff
@@ -325,31 +404,55 @@ def _on_chamber_state(self, state):
...
+                        self.im_opt = insert_tile_to_image(opt, self.im_opt)
+                    if sem is not None:
+                        self.im_sem = insert_tile_to_image(sem, self.im_sem)
+
+                    # Merge optical and sem overview images
+                    self.ovv_im = merge_opt_sem(self.im_opt, self.im_sem)
+
+                    # Update display
+                    if self.upd_stream:
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: Why would `self.upd_stream` be equivalent to False? It seems this `if` block is always executed.

Moreover, I'd recommend to only run it once per update (at most). Maybe you could just check `.curr_s` instead of looking at every stream (which should be doing the same thing as long as not 2 streams are playing simultaneously). Or you'd need to check if ovv_im has been updated.

File: src/odemis/gui/cont/views.py [hunk 5]
```diff
@@ -317,6 +370,32 @@ def on_stage_pos_change(self, p_pos):
...
+
+        if self.curr_s:
+            self._update_ovv()
+            self.curr_s.should_update.subscribe(self._on_stream_update, init=True)
+            self.curr_s.image.subscribe(self._onNewImage)
+
+    def _on_stream_update(self, _):
+        self._update_ovv()
+
+    @wxlimit_invocation(1)  # max 1 Hz
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: Just use `@limit_invocation()`, the `wx` variant ensures also that you're running in the main GUI thread, but it's not needed here.

File: src/odemis/gui/cont/views.py [hunk 6]
```diff
@@ -317,6 +370,32 @@ def on_stage_pos_change(self, p_pos):
...
+            self.curr_s = streams[0]
+        except IndexError:
+            self.curr_s = None
+
+        if self.curr_s:
+            self._update_ovv()
+            self.curr_s.should_update.subscribe(self._on_stream_update, init=True)
+            self.curr_s.image.subscribe(self._onNewImage)
+
+    def _on_stream_update(self, _):
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: What's the point of updating the image when the stream starts or stop playing? Anyway, you're already getting called at every image update of the stream.

**Comment by philipwinklerdelmic** [comment]
I made the changes you suggested. 

I still see the error message (strangely now more than before making the changes).

17:31:14 ERROR   log            : Subscriber <odemis.util.weak.WeakMethodBound object at 0x7f2f538609d0> raised exception when receiving value 2.0090705927e-05
Traceback (most recent call last):
  File "/home/winkler/development/odemis/src/odemis/model/_vattributes.py", line 101, in notify
    l(v)
  File "/home/winkler/development/odemis/src/odemis/util/weak.py", line 37, in __call__
    return self.f(ins, *arg, **kwargs)
  File "<decorator-gen-90>", line 2, in _on_view_mpp
  File "/home/winkler/development/odemis/src/odemis/gui/util/__init__.py", line 49, in call_in_wx_main
    return f(self, *args, **kwargs)
  File "/home/winkler/development/odemis/src/odemis/gui/comp/viewport.py", line 357, in _on_view_mpp
    self.microscope_view.fov.value = fov
  File "/home/winkler/development/odemis/src/odemis/model/_vattributes.py", line 1107, in _set_value
    VigilantAttribute._set_value(self, value, **kwargs)
  File "/home/winkler/development/odemis/src/odemis/model/_vattributes.py", line 201, in _set_value
    self._check(value) # we allow the setter to even put illegal value, it's the master
  File "/home/winkler/development/odemis/src/odemis/model/_vattributes.py", line 1116, in _check
    Continuous._check(self, value)
  File "/home/winkler/development/odemis/src/odemis/model/_vattributes.py", line 902, in _check
    if not isinstance(value, collections.Iterable):
  File "/usr/lib/python2.7/abc.py", line 132, in __instancecheck__
    if subclass is not None and subclass in cls._abc_cache:
RuntimeError: maximum recursion depth exceeded

File: src/odemis/gui/cont/views.py [hunk 7]
```diff
@@ -285,22 +290,74 @@ def focusViewWithStream(self, stream):
...
+                ovv_shape = (2 * int(max_y / self.m_view.mpp.value),
+                             2 * int(max_x / self.m_view.mpp.value), 3)
+            else:
+                ovv_shape = (int(OVV_DIMENSION[1] / self.m_view.mpp.value),
+                             int(OVV_DIMENSION[0] / self.m_view.mpp.value), 3)
+        else:
+            ovv_shape = (int(OVV_DIMENSION[1] / self.m_view.mpp.value),
+                         int(OVV_DIMENSION[0] / self.m_view.mpp.value), 3)
+
+        self.ovv_im = numpy.zeros(ovv_shape)
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: Numpy defaults to creating arrays as floats, but you can directly override the dtype. So the two lines can be simplied by:
`self.ovv_im = numpy.zeros(ovv_shape, dtype=numpy.uint8)`

File: src/odemis/gui/cont/views.py [hunk 8]
```diff
@@ -285,22 +290,74 @@ def focusViewWithStream(self, stream):
...
+            self.m_view.addStream(overview_stream)
+            # TODO: add it to self.tab_data_model.streams?
+
+        # Built-up overview image
+        # Initialize the size of the ovv image with the stage size if the stage is small (< 5cm),
+        # otherwise fall back to OVV_DIMENSION
+        ax_x = self.main_data.stage.axes["x"]
+        ax_y = self.main_data.stage.axes["y"]
+        if (hasattr(ax_x, "range") and hasattr(ax_y, "range")):
+            max_x = max(numpy.absolute(ax_x.range))
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: The range can be something like [1.1, 1.2], in which case it'll have a 0.1m "range width". You actually only want to know the width of the range. => use `ax_x[1] - ax_x[0]`.

File: src/odemis/gui/cont/views.py [hunk 9]
```diff
@@ -325,31 +407,48 @@ def _on_chamber_state(self, state):
...
+        if self._data_model.streams.value and self.curr_s \
+                    and self.curr_s.image.value is not None:
+            s = self.curr_s
+            img = s.image.value
+            if isinstance(s, acqstream.OpticalStream):
+                self.im_opt = insert_tile_to_image(img, self.im_opt)
+            elif isinstance(s, acqstream.EMStream):
+                self.im_sem = insert_tile_to_image(img, self.im_opt)
+            else:
+                logging.info("Stream %s not added to overview image" % (s))
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: `logging` supports passing the inputs for the string formatting as extra arguments, and it's the recommended way. So you can write:
`logging.info("%s not added to overview image as it's not optical nor EM", s)`

File: src/odemis/gui/cont/views.py [hunk 10]
```diff
@@ -325,31 +407,48 @@ def _on_chamber_state(self, state):
             # sample have probably nothing in common with this new sample
             self._data_model.stage_history.value = self._data_model.stage_history.value[-1:]
 
+    def _update_ovv(self):
+        """ Update the overview image with the currently active stream. """
+        if self._data_model.streams.value and self.curr_s \
```
Comment (pieleric) [inline src/odemis/gui/cont/views.py]: Is it useful to check `self._data_model.streams.value` is not empty? If it's empty, `curr_s` should be `None`. Or are there special cases? In that case add a comment.

File: src/odemis/gui/util/img.py [hunk 11]
```diff
@@ -2098,3 +2099,109 @@ def wxImageScaleKeepRatio(im, size, quality=wx.IMAGE_QUALITY_NORMAL):
...
+    """ 
+    Merges two images (im and background) into one.
+    im, background: DataArray with opt.shape = background.shape (either RGB or RGBA)
+    returns RGB DataArray
+    """
+    assert(im.shape == background.shape, "Images have different shapes.")
+    if im.shape[-1] != 3 and im.shape[-1] != 4:
+        raise ValueError("Ovv images have an invalid number of channels: %d" % (im.shape[-1]))
+
+    im = copy.deepcopy(im)
```
Comment (pieleric) [inline src/odemis/gui/util/img.py]: deepcopy is very expensive, as it'll allocate a new memory area, to duplicate the array. You probably don't need this. If the goal is just not modify the original metadata, you can just create a new DataArray with a (deep)copy of the metadata.

File: src/odemis/gui/util/img.py [hunk 12]
```diff
@@ -2098,3 +2099,109 @@ def wxImageScaleKeepRatio(im, size, quality=wx.IMAGE_QUALITY_NORMAL):
...
+    # Combine images
+    surface = cairo.ImageSurface.create_for_data(
+        background, cairo.FORMAT_ARGB32, buffer_size[0], buffer_size[1])
+    ctx = cairo.Context(surface)
+
+    draw_image(ctx, im, buffer_top_left, buffer_top_left, buffer_scale,
+               buffer_size, 1, blend_mode=BLEND_SCREEN)
+
+    # Convert back to RGB
+    rgba = numpy.empty((im.shape[0], im.shape[1], 3), dtype=numpy.uint8)
```
Comment (pieleric) [inline src/odemis/gui/util/img.py]: Update the name of the variable to `rgb`, as it's only RGB.

File: src/odemis/acq/stream/_static.py [hunk 13]
```diff
@@ -935,3 +935,52 @@ def onSpectrumBandwidth(self, value):
         """
         self._updateHistogram()
         self._shouldUpdateImage()
+
+
+class RGBUpdatableStream(StaticStream):
```
[the remark this answers is not in the document]
Comment (philipwinklerdelmic) [reply src/odemis/acq/stream/_static.py]: I inherit from StaticStream because the functions _init_projection_vas and _init_thread are overwritten with an empty function in RGBStream, but these functions are needed to properly update the image.

File: src/odemis/gui/util/img.py [hunk 14]
```diff
@@ -2098,3 +2099,107 @@ def wxImageScaleKeepRatio(im, size, quality=wx.IMAGE_QUALITY_NORMAL):
...
+    """ 
+    Merges two images (im and background) into one.
+    im, background: DataArray with opt.shape = background.shape (either RGB or RGBA)
+    returns RGB DataArray
+    """
+    assert(im.shape == background.shape, "Images have different shapes.")
+    if im.shape[-1] != 3 and im.shape[-1] != 4:
+        raise ValueError("Ovv images have an invalid number of channels: %d" % (im.shape[-1]))
+
+    md1 = copy.deepcopy(im.metadata)
```
Comment (pieleric) [inline src/odemis/gui/util/img.py]: You are now not using `md1` and `md2` at all. You need to also change im into a new DataArray with that new metadata. You can do it with this line:
`im = model.DataArray(im, md1)`

BTW, it seems that `dc_keepalpha` is not useful for `background`, so you could avoid duplicating the metadata.

File: src/odemis/gui/util/img.py [hunk 15]
```diff
@@ -2098,3 +2099,107 @@ def wxImageScaleKeepRatio(im, size, quality=wx.IMAGE_QUALITY_NORMAL):
...
+    md2 = copy.deepcopy(background.metadata)
+    md1['dc_keepalpha'] = True
+    md2['dc_keepalpha'] = True
+    im = format_rgba_darray(im, 255)  # convert to BGRA
+    background = format_rgba_darray(background, 255)
```
Comment (pieleric) [inline src/odemis/gui/util/img.py]: Actually, now that you directly use `background` as the cairo surface, it should be duplicated.

File: src/odemis/gui/util/test/img_test.py [hunk 16]
```diff
@@ -601,6 +601,80 @@ def test_first(self):
...
+        ovv_im = model.DataArray(numpy.zeros((5, 5, 3)))
+        ovv_im.metadata[model.MD_PIXEL_SIZE] = (1, 1)
+        tile = model.DataArray(255 * numpy.ones((2, 2, 3)))
+        tile.metadata[model.MD_POS] = (10, -10)
+        tile.metadata[model.MD_PIXEL_SIZE] = (1, 1)
+        ovv_im_new = insert_tile_to_image(tile, ovv_im)
+        numpy.testing.assert_array_equal(ovv_im_new, numpy.zeros((5, 5, 3)))
+
+    def test_merge(self):
+        """ Tests merge_opt_sem function """
```
Comment (pieleric) [inline src/odemis/gui/util/test/img_test.py]: That comment should be updated with the new name of the function.

File: src/odemis/gui/util/test/img_test.py [hunk 17]
```diff
@@ -601,6 +601,80 @@ def test_first(self):
...
+            model.MD_EXP_TIME: 1,  # s
+            model.MD_IN_WL: (600e-9, 620e-9),  # m
+            model.MD_OUT_WL: (620e-9, 650e-9),  # m
+            model.MD_USER_TINT: (0, 0, 255),  # RGB (blue)
+            model.MD_ROTATION: 0.1,  # rad
+            model.MD_SHEAR: 0,
+            model.MD_DIMS: "YXC"
+        }
+
+        opt = model.DataArray(200 * numpy.ones((10, 10, 3)), md)
```
Comment (pieleric) [inline src/odemis/gui/util/test/img_test.py]: Please use `uint8` as dtype, to test more realistic images.

Also add a test case when the images have alpha channel (completely fill-up is fine), as it's supposed to be supported.

File: src/odemis/gui/util/test/img_test.py [hunk 18]
```diff
@@ -601,6 +601,80 @@ def test_first(self):
...
+        sem = model.DataArray(10 * numpy.ones((10, 10, 3)), md)
+        merged = merge_screen(opt, sem)
+        numpy.testing.assert_equal(True, numpy.any(numpy.not_equal(merged, opt)))
```
Comment (pieleric) [inline src/odemis/gui/util/test/img_test.py]: What does this tests for? It's really hard to follow the logic of this line. It checks that at least one pixel is different? In that case, replacing by just `assert_not_equal(merged, opt)` would make the line clearer.

Comment (philipwinklerdelmic) [reply src/odemis/gui/util/test/img_test.py]: To my knowledge, there doesn't exist an assert_not_equal function for arrays. I changed the code now to 
        self.assertRaises(AssertionError, numpy.testing.assert_array_equal,
                                    merged, opt)

I hope that's easier to read.

File: src/odemis/gui/util/img.py (hunk 15 above)
Comment (philipwinklerdelmic) [reply src/odemis/gui/util/img.py]: Are the metadata changed when using cairo as background? The test also passes if I don't deepcopy it.
